python/paddle/fluid/dygraph/parallel.py

```diff
@@ -58,3 +65,37 @@ def dev_id(self):
     @property
     def current_endpoint(self):
         return self._current_endpoint
+
+    @property
+    def trainer_endpoints(self):
+        return self._trainer_endpoints
+
+
+class DataParallel(layers.Layer):
+    def __init__(self, layers):
+        super(DataParallel,
+              self).__init__(layers.full_name() + "_data_parallel")
+        self._layers = layers
+
+    def build_once(self, *inputs, **kwargs):
+        #TODO(Yancey1989): broadcast all the paramters
+        pass
+
+    def forward(self, *inputs, **kwargs):
+        def _collective_hook(iop):
+            op = _dygraph_tracer()._ops[iop._trace_id]
+            for k, v in six.iteritems(op.inputs):
+                for ivar in v:
+                    g = ivar._grad_ivar()
+                    g_var = Variable(
+                        block=self._helper.main_program.current_block(),
+                        name=ivar._grad_name(),
+                        stop_gradient=True,
+                        ivar=g)
+                    collective._allreduce(g_var, g_var, sync_mode=True)
+
+        outs = self._layers(*inputs, **kwargs)
+        for _, op in six.iteritems(_dygraph_tracer()._ops):
```

Comment:
should all op's output be collectived?

Reply:
Sure, we need to knew CPUPlace or CUDAPlace event the GPU index.